File: vulns/Crypto/samples/C/vulnerable.c
// Vuln 1
// Clave literal
// ECB manual
// Sin IV ni auth

#include <openssl/aes.h>
#include <string.h>

/* Clave incrustada: recuperable con `strings` o desensamblado */
static const unsigned char KEY[16] = "hardcoded_key_16";

void encrypt_note(const unsigned char *in, unsigned char *out, size_t nblocks) {
    AES_KEY aes;
    AES_set_encrypt_key(KEY, 128, &aes);
    for (size_t i = 0; i < nblocks; i++)
        AES_encrypt(in + i*16, out + i*16, &aes);   /* ECB: bloque igual -> ct igual */
}

// Vuln 2
// memcmp/strcmp sobre secretos
// 

#include <string.h>

static const char *ADMIN_TOKEN = "s3cr3t_admin_token";

int is_admin(const char *supplied) {
    /* memcmp retorna en el primer byte diferente -> fuga temporal */
    return memcmp(supplied, ADMIN_TOKEN, strlen(ADMIN_TOKEN)) == 0;
}

File: vulns/Crypto/samples/JS/vulnerable.js
// Vuln 1
// Clave literal en el codigo
// aes-256-ecb -> ECB byte at a time
// Sin IV ni auth

const crypto = require("crypto");

const KEY = Buffer.from("0123456789abcdef0123456789abcdef"); // 32 bytes hardcodeados

function encryptNote(plaintext) {
  const cipher = crypto.createCipheriv("aes-256-ecb", KEY, null); // ECB
  return Buffer.concat([cipher.update(plaintext), cipher.final()]);
}

// Vuln 2
// Math.random() para tokens, IDs de sesión, OTP, etc.
// Math.random es predecible en V8
// A veces mezclado con Date.now() como "entropía".

function makeToken(len = 32) {
  let t = "";
  const cs = "abcdef0123456789";
  for (let i = 0; i < len; i++)
    t += cs[Math.floor(Math.random() * cs.length)]; // PRNG no seguro
  return t;
}

// Vuln 3
// === / == sobre secretos, tokens o firmas HMAC.
// Diferencias de tiempo según cuántos bytes coinciden.

const API_KEY = "s3cr3t_admin_key";

function isAdmin(supplied) {
  return supplied === API_KEY;   // comparación corto-circuito -> timing
}

// Vuln 4
// jwt.verify sin la opción algorithms
// Secreto corto/adivinable
// algorithm confusion

const jwt = require("jsonwebtoken");

const SECRET = "secret"; // débil + hardcodeado

function verifyToken(token) {
  return jwt.verify(token, SECRET);
}